Rule improvement: ReturnFromFinally

rwqwr opened the pull request:
Fixes https://github.com/detekt/detekt/issues/3197

schalkms reviewed (changes requested):
Thanks for tackling this issue! 🙂
It's a really nice 1st contribution to detekt. 

Please note, that there's already an existing rule. The following rule should be adapted to handle this case. There's no need for a new rule. Otherwise, this is very confusing for users of detekt.
Please take a look at the following class.
https://github.com/detekt/detekt/blob/master/detekt-rules-exceptions/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/exceptions/ReturnFromFinally.kt

Please kindly let me know, if you have any questions.

rwqwr commented:
Thank you for your response! 🙂
I fixed this moment and implemented this case in `ReturnFromFinally` rule.

schalkms commented on detekt-rules-exceptions/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/exceptions/ReturnFromFinally.kt:
```diff
@@ -10,12 +10,18 @@ import io.gitlab.arturbosch.detekt.api.Severity
+import org.jetbrains.kotlin.resolve.calls.callUtil.getType
 
 /**
  * Reports all `return` statements in `finally` blocks.
  * Using `return` statements in `finally` blocks can discard and hide exceptions that are thrown in the `try` block.
+ * Also report unaffected return value in `finally` block if it used as expression.
+ * See: https://kotlinlang.org/docs/reference/exceptions.html#try-is-an-expression
```
This might be better understandable to first time users of this rule.

```suggestion
 * Reports all `return` statements in `finally` blocks.
 * Using `return` statements in `finally` blocks can discard and hide exceptions that are thrown in the `try` block.
 * 
 * Furthermore, this rule reports values from `finally` blocks, if the corresponding `try` is used as an expression.
```

schalkms commented on detekt-rules-exceptions/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/exceptions/ReturnFromFinally.kt:
```diff
@@ -45,6 +54,30 @@ class ReturnFromFinally(config: Config = Config.empty) : Rule(config) {
+
+        val finallyBlock = expression.finallyBlock
+
+        if (expression.isUsedAsExpression(bindingContext) && finallyBlock != null) {
```
This statement can be switched, since the usage of `bindingContext` in this case could be more expensive.

```suggestion
        if (finallyBlock != null && expression.isUsedAsExpression(bindingContext)) {
```

schalkms commented on detekt-rules-exceptions/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/exceptions/ReturnFromFinally.kt:
```diff
@@ -45,6 +54,30 @@ class ReturnFromFinally(config: Config = Config.empty) : Rule(config) {
             .forEach { report(CodeSmell(issue, Entity.from(it), issue.description)) }
     }
```
Maybe the logic from `visitFinallySection()` can
* be merged together with `visitTryExpression()`
* use the `bindingContext` capability as well.

This would reduce the complexity of this rule.

schalkms reviewed (commented):
Thanks for incorporating the changes! This PR looks good.
The mentioned refactoring can also be done in a separate PR, if you want.

rwqwr commented:
I refactored code and changed the implementation of `visitFinallySection()` with the `bindingContext` capability, but left the check inside it. It seemed to me more correct.

schalkms commented on detekt-rules-exceptions/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/exceptions/ReturnFromFinally.kt:
```diff
@@ -66,16 +69,25 @@ class ReturnFromFinally(config: Config = Config.empty) : Rule(config) {
+
+        val targetFunctionBodyExpressionStatements = targetFunction
+            .blockExpressionsOrSingle()
+            .toList()
```
Is this mandatory?

rwqwr replied:
`blockExpressionsOrSingle()` extension return `Sequence<KtElement>` and `isInsideOf` take `Iterable<PsiElement>` as argument.

schalkms commented on detekt-rules-exceptions/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/exceptions/ReturnFromFinally.kt:
```diff
@@ -66,16 +69,25 @@ class ReturnFromFinally(config: Config = Config.empty) : Rule(config) {
+        blockExpression: KtBlockExpression,
+        returnStmts: KtReturnExpression
+    ): Boolean {
+        val targetFunction = returnStmts.getTargetFunction(bindingContext)
```
I guess we also need to check for an empty binding context here.

rwqwr replied:
I added check to `visitFinallySection()`
Is it ok?

schalkms commented on detekt-rules-exceptions/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/exceptions/ReturnFromFinallySpec.kt:
```diff
@@ -19,16 +18,29 @@ class ReturnFromFinallySpec : Spek({
             val code = """
-            fun x() {
+                class Main {
+                fun main() {
+                Second().main()
+                }
+                }
+                
+                class Second {
+                fun main() {
+                test()
+                }
+                
+                fun test() {
                 try {
                 } finally {
                     return
                 }
-            }
+                return
+                }
+                }
```
Unfortunately, the formatting seems to be off here.

rwqwr replied:
Sorry, it is my mistake :)

schalkms commented:
Thanks! We also need to add a `@requireTypeResolution` annotation to the rule comment, since type and symbol solving needs to be configured.

rwqwr commented on detekt-rules-exceptions/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/exceptions/ReturnFromFinally.kt:
```diff
@@ -25,9 +31,12 @@ import org.jetbrains.kotlin.psi.psiUtil.parents
  * </noncompliant>
  *
  * @configuration ignoreLabeled - ignores labeled return statements (default: `false`)
+ * @requiresTypeResolution
```
> We also need to add a `@requireTypeResolution` annotation to the rule comment, since type and symbol solving needs to be configured.

Do you mean this annotation?

schalkms replied:
Exactly
